Added interpolators

wleroux opened the pull request:
Hello,

I noticed that Mini2Dx has a great emphasis on the interpolation step, however aside from the basic interpolation provided by Mini2Dx, it's a bit left to the developer's imagination. Additionally, the 'previous value' changes when you lerp it on the current implementation (Ow!) For instance, the following fails:

```
    @Test
    public void testLerping() {
        CollisionBox box = new CollisionBox();
        box.set(1,2,3,4);
        box.rotateAround(new Point(22, 55), 22);

        box.preUpdate();;
        box.set(5,6,7,8);
        box.rotateAround(new Point(55, 22), -22);

        box.interpolate(null, 0.55f);
        float renderX = box.getRenderX();
        float renderY = box.getRenderY();

        box.interpolate(null, 0.55f);
        Assert.assertEquals(renderX, box.getRenderX());
        Assert.assertEquals(renderY, box.getRenderY());
    }
```

I understand the 'mucking' with the values is to reduce the object allocation, so I built an interpolation system that is both easy to implement new types, doesn't muck with your values, and supports the flyweight approach.

Here's a sample of usage with primitives (i.e.: no flyweights):

```
  @Test
  public void testInterpolatedValues() {
    InterpolatedFloatValue value = new InterpolatedFloatValue(Interpolators.linearFloat());
    value.setInitialValue(initialValue);
    value.setTargetValue(targetValue);
    for (float alpha = 0.0f; alpha <= 1.0f; alpha += 0.05f) {
      value.setAlpha(alpha);
      Assert.assertEquals(initialValue + alpha * (targetValue - initialValue), value.getInterpolatedValue(), 0.0f);
    }
  }
```

Here's a sample of usage with non-primitives:

```
  @Test
  public void testInterpolateValues() {
    interpolatedValue.setInitialValue(initialValue);
    interpolatedValue.setTargetValue(targetValue);

    Circle flyweight = new Circle(0f, 0f, 0f);
    for (float alpha = 0.0f; alpha <= 1.0f; alpha += 0.05f) {
      interpolatedValue.setAlpha(alpha);

      float x = initialValue.getX() + alpha * (targetValue.getX() - initialValue.getX());
      float y = initialValue.getY() + alpha * (targetValue.getY() - initialValue.getY());
      float radius = initialValue.getRadius() + alpha * (targetValue.getRadius() - initialValue.getRadius());

      interpolatedValue.getInterpolatedValue(flyweight);
      Assert.assertEquals("x at " + alpha, x, flyweight.getX(), 0.05f);
      Assert.assertEquals("y at "+ alpha, y, flyweight.getY(), 0.05f);
      Assert.assertEquals("radius at "+ alpha, radius, flyweight.getRadius(), 0.05f);
    }
  }
```

Currently, the InterpolatedValue class has a 'initial value' and 'target value' which may be a bit strange. Maybe 'initial value' and 'current value' makes more sense?

In any case, I haven't dabbled in the serialization aspects of it (CollisionCircle doesn't seem to have serialization-specific details.) Maybe implements Serializable is enough?

A side note, I couldn't replicate the CollisionBox linear interpolation. The Rectangle class under the hood does some wicked things with the rotation. As soon as I add in linear interpolation of a moving rotational center, the x-y coordinates get 'rotated' and I lose my 'initial values'. For the time being, I left it out. CollisionBox uses some private methods to set up the interpolated value.

Thanks! I appreciate your thoughts on the matter.

wleroux commented on core/src/main/java/org/mini2Dx/core/geom/Circle.java:
```diff
@@ -29,7 +29,7 @@ public Circle(int radius) {
 		this.radius = radius;
 	}
 	
-	public Circle(float centerX, float centerY, int radius) {
```
This is an important (and unrelated) change to the rest of the commit. When working with this, I noticed that the circle had an 'int' argument and a 'float' variable backing it up.

wleroux commented on core/src/main/java/org/mini2Dx/core/engine/interpolator/geom/LinearCircleInterpolator.java:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  @Override
+  public Circle interpolate(Circle flyweight, Circle initialValue, Circle targetValue, float alpha) {
```
This class shows how more abstract classes can use the interpolators recursively; the circle has 3 floats that needs to be interpolated, but we don't want to re-write that interpolation logic everywhere.

tomcashman commented:
I like this idea. Some thoughts:

* Should Interpolator#interpolate be void since the flyweight is passed anyway?
* I agree that 'target value' sounds strange, let's go with 'current value'
* Could you add the [copyright notice](https://github.com/mini2Dx/mini2Dx/blob/master/core/src/main/java/org/mini2Dx/core/engine/Positionable.java) to the top of all new files? - I should add this point to the contributing guide.

Yeah the Rectangle class is not the cleanest code. It underwent a lot of changes and rewrites over time as I found performance bottlenecks when I was implementing my last game. The good news is that it's fast, the bad news is its difficult to maintain. Hence there's a lot of unit tests for it :stuck_out_tongue:

wleroux commented:
Glad you liked the idea! I agree on all counts.

Is the Rectangle class tested for performance or implementation correctness only? I noticed there were no tests for rotation when I was looking at the tests for it. I'm considering having a stab at it.

In regards for `Interpolator#interpolate`, there are benefits to having the flyweight be returned by the function, however I believe they do not outweigh the costs. I'll list them here in case someone looks up why we did it this way:

The benefits to having the return value as both an argument and return value:

1. It is a 'more' consistent interface as primitives (primitives always return the values)
2. While the flyweight is an argument, it is not necessarily required. (i.e.: the function doesn't use the flyweight and instantiates a new object as the return value.)

The drawbacks are:

1. The user of the API is never sure to use the flyweight or the returned value
2. The flyweight and the returned value could be different depending on the implementation
3. By using return values, it increases the likelihood of object allocation (which should definitely be avoided in the interpolation step.)

By far, having a consistent and easily understood API that mitigates a tendency to allocate objects is a better choice for the framework... Despite primitives having a slightly different pattern.

wleroux commented:
In this new version, I've moved the alpha into the `InterpolatedValue#getInterpolatedValue(T flyweight, float alpha);` function. I'm not sure whether we want to do this during the 'interpolate' phase (where we'd set world.setAlpha(alpha);` or the render phase. The way I see it, it would require one less iteration of entities to store the same alpha everywhere: the render phase knows the alpha (from world) and can apply it when generating the interpolated values using the flyweights.

I've added some documentation to the JavaDoc as well.

tomcashman commented:
The only argument I'd have against moving it to the render phase is that this way it's more explicit and allows the developer to remove components during the interpolate phase so that an entity is not included in the render phase.

Anyway everything looks good so I'll accept the merge request if you're happy with everything?

wleroux commented:
I think I prefer it the way it is...

You can always filter in the Interpolation phase:

```
value.getInterpolatedValue(flyweight, alpha);
```
Instead of:
```
value.setAlpha(alpha);
value.getInterpolatedValue(flyweight);
```

If you just use the alpha, then you don't need to get the interpolated value at all.

Additionally, it prevents the case of someone forgetting to set the alpha; by asking for both, we are explicit about the temporal coupling. I've already forgotten to set the alpha in my test case and was wondering why my expectations were not being met.
